Add three test cases for Evolution

QingmingSu opened the pull request:
evolution_smoke: Case #1503857: Evolution setup assistant
evolution_mail_imap: Case #1503768: send and receive email via IMAP
evolution_mail_ews: Case #1503965: Setup MS Exchange account

coolo commented:
please add a copyright and license header, e.g.

```
# Evolution tests
#
# Copyright © 2016 SUSE LLC
#
# Copying and distribution of this file, with or without modification,
# are permitted in any medium without royalty provided the copyright
# notice and this notice are preserved.  This file is offered as-is,
# without any warranty.
```

GPL header will do too, but is longer ;)

okurz commented on tests/x11regressions/evolution/evolution_mail_ews.pm:
```diff
@@ -0,0 +1,128 @@
+    }
+
+    # Follow the wizard to setup mail account
+    assert_screen "test-evolution-1";
```
the following should work faster as the test would not need to wait for timeout on check_screen:
```
assert_screen([qw/evolution-default-client-ask test-evolution-1/]);
if (match_has_tag('evolution-default-client-ask')) {
    assert_and_click 'evolution-default-client-agree';
    assert_screen 'test-evolution-1';
}
```

and the same some lines down but both approaches work. only change it if you like it.

QingmingSu replied:
Hi okurz,

Thanks for the recommendation.  I tried to replace the api check_screen with match_has_tag, but the test script didn't work as expected. After digging into the definition of match_has_tag function, I find it is improper to be used here. Because the precondition of using match_has_tag is that there must be a matched needle previously, and then we can judge if there is a wanted tag in the matched needle. But here, I want to match a tag from a new screen (which is not 100% sure to pop up), so match_has_tag doesn't work in this case.

QingmingSu replied:
Seems I didn't understand your suggestion well when I wrote my previous note. You suggested to match multiple possible needles in one assert_screen and handle the rare case in if condition, which would make the test faster in most cases.
Thanks a lot for the approach.

okurz commented on tests/x11regressions/evolution/evolution_mail_imap.pm:
```diff
@@ -0,0 +1,192 @@
+    my $mail_server   = 'mail.microfocus.com';
+    my $mail_user     = 'zzzSUSEExTest19';
+    my $mail_passwd   = 'P@$$w0rd2015';
+    my $mail_revport  = '993';
```
should say "mail_recvport"

okurz commented:
-1

see some inline notes. Also you are adding three files which are very similar. I strongly recommend to extract common code and reduce the duplication. Also see [DRY](http://programmer.97things.oreilly.com/wiki/index.php/Don't_Repeat_Yourself)

coolo commented:
we can do the refactoring once the tests are in and work. (who talked about scaring away contributors?)

QingmingSu commented:
Hi coolo,

I have added the copyright for all the three evolution cases and rebased my local master branch upon the upstream.
The updated commit has been pushed to the forked repository. And the cases have been validated by rerunning before committed.

Thanks for the reminder,
Qingming

QingmingSu commented:
Hi okurz,

I'd like to reduce the duplication by sharing the common codes between test cases. But I didn't find an example to make it. 
Is there any example to share the same codes among several test cases? 

Thanks,
Qingming

QingmingSu commented:
Add a note to explain why the internal microfocus account is used in the test scripts. At the beginning, I tried to use some gmail or hotmail accounts in the Evolution testing (from wiki https://twiki.innerweb.novell.com/bin/view/Main/ChatSoftwareTest), but most of the accounts are invalid now. The gmail and hotmail accounts must be activated again if it isn't accessed within some time. But it is impossible to re-activate the accounts for some reasons, such as the guy who created the account had left the company.

okurz commented:
> I'd like to reduce the duplication by sharing the common codes between test cases. But I didn't find an example to make it. Is there any example to share the same codes among several test cases?

I must admit I did not try this myself as I did not write new test cases so far, only adapted existing ones but `lib/x11test.pm` should be a good example by itself. You can either put helper methods in there or define a new baseclass which uses x11test as base and use that. I recommend the second option.

you could try that in your next tests or as a refactoring of the current tests.

coolo commented:
you can put a common base class into lib like okurz describes - or you can put a base class in the test directory itself like we did with tests/qa_automation/qa_run.pm

QingmingSu commented:
Got it, thanks.
